feat: gamestore cpp

murilo09 opened the pull request:
# Description

Moving Gamestore to CPP

## Type of change

  - [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

  - [ ] My code follows the style guidelines of this project
  - [ ] I have performed a self-review of my own code
  - [ ] I checked the PR checks reports
  - [ ] I have commented my code, particularly in hard-to-understand areas
  - [ ] I have made corresponding changes to the documentation
  - [ ] My changes generate no new warnings
  - [ ] I have added tests that prove my fix is effective or that my feature works

murilo09 committed:
feat: gamestore cpp

murilo09 committed:
fix: store history

murilo09 committed:
fix: using shared pointer

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
fix: build

murilo09 committed:
fix: tibia coins balance

murilo09 committed:
fix: more descriptive log

murilo09 committed:
fix: store history

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
fix: only wrapable store items can be moved to a house

murilo09 committed:
feat: movable store items flag

murilo09 committed:
fix: duplicated house store items

murilo09 committed:
fix: update balance after purchase

murilo09 committed:
fix: saving premium offer on db

phacUFPE reviewed (commented):
Here's a list of the missing contents.

- [x] Store only accepting transferable tibia coins as payment
- [ ] Bundle
- [ ] Hirelings
- [X] Name change
- [X] XP Boost
- [ ] Check if Vip System activated
- [ ] Integration with Cyclopedia info and history
- [x] Change string format to fmt

phacUFPE commented on src/io/functions/iologindata_load_player.cpp:
```diff
@@ -828,6 +828,27 @@ void IOLoginDataLoad::loadPlayerForgeHistory(std::shared_ptr<Player> player, DBR
+	}
+
+	std::ostringstream query;
+	query << "SELECT * FROM `store_history` WHERE `account_id` = " << player->getAccountId();
```
Please prefer using fmt as stringstream.
```suggestion
	const std::string query = fmt::format("SELECT * FROM `store_history` WHERE `account_id` = {}", player->getAccountId());
```

phacUFPE commented on src/creatures/players/player.cpp:
```diff
@@ -7749,6 +7750,128 @@ void Player::closeAllExternalContainers() {
 	}
 }
 
+bool Player::canBuyStoreOffer(const Offer* offer) {
```
I think that is better using smart pointers.

murilo09 replied:
Let's focus on making everything work for now and address these changes later.

phacUFPE commented on src/game/game.cpp:
```diff
@@ -10282,20 +10671,134 @@ bool Game::addItemStoreInbox(std::shared_ptr<Player> player, uint32_t itemId) {
+
+	std::ostringstream query;
+	Database &db = Database::getInstance();
+	query << "SELECT `id` FROM `player` WHERE `name` = " << db.escapeString(newName);
```
Please prefer using fmt over ostring.
```suggestion
	Database &db = Database::getInstance();
	std::string query = fmt::format("SELECT `id` FROM `players` WHERE `name` = {}", db.escapeString(newName));
```

phacUFPE commented on src/game/game.cpp:
```diff
@@ -10282,20 +10671,134 @@ bool Game::addItemStoreInbox(std::shared_ptr<Player> player, uint32_t itemId) {
+	}
+
+	capitalizeWords(newName);
+	query << "UPDATE `players` SET `name` = " << db.escapeString(newName) << "WHERE `id` = " << player->getGUID();
```
```suggestion
	query = fmt::format("UPDATE `players` SET `name` = {} WHERE `id` = {}", db.escapeString(newName), player->getGUID());
```

phacUFPE commented on src/io/functions/iologindata_save_player.cpp:
```diff
@@ -720,6 +720,42 @@ bool IOLoginDataSave::savePlayerForgeHistory(std::shared_ptr<Player> player) {
+
+	if (!insertQuery.execute()) {
+		return false;
+	}
```
```suggestion
	std::string query = fmt::format("DELETE FROM `store_history` WHERE `account_id` = {}", player->getAccountId());
	if (!Database::getInstance().executeQuery(query.str())) {
		return false;
	}

	DBInsert insertQuery("INSERT INTO `store_history` (`account_id`, `description`, `coin_amount`, `coin_type`, `type`, `time`) VALUES");
	for (const auto &historyEntry : player->getStoreHistory()) {
		const auto descriptionString = Database::getInstance().escapeString(historyEntry.description);
		// Append query informations
		query = fmt::format("{}, {}, {}, {}, {}, {}", player->getAccountId(), descriptionString, historyEntry.coinAmount, historyEntry.coinType, historyEntry.historyType, historyEntry.createdAt);

		if (!insertQuery.addRow(query)) {
			return false;
		}
	}

	if (!insertQuery.execute()) {
		return false;
	}
```

murilo09 committed:
fix: account_coins enum values

murilo09 committed:
fix: build

murilo09 committed:
fix: store transfer history

murilo09 committed:
fix: xp boost offer

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
fix: migration version

dudantas commented on data/scripts/globalevents/server_initialization.lua:
```diff
@@ -10,7 +10,7 @@ local function cleanupDatabase()
 
 	db.query("UPDATE `players` SET `isreward` = " .. DAILY_REWARD_NOTCOLLECTED)
-	db.query("UPDATE `player_storage` SET `value` = 0 WHERE `player_storage`.`key` = 51052")
+	db.query("UPDATE `player_storage` SET `value` = 1 WHERE `player_storage`.`key` = 51052")
```
Did it change by mistake or was there some change regarding this?

dudantas commented on src/account/account_repository.hpp:
```diff
@@ -38,6 +38,15 @@ class AccountRepository {
+	virtual bool registerStoreTransaction(
+		const uint32_t &id,
+		uint8_t type,
+		int32_t amount,
```
Here it needs to be changed to uint32_t

dudantas commented on src/account/account_repository_db.cpp:
```diff
@@ -148,6 +148,42 @@ bool AccountRepositoryDB::registerCoinsTransaction(
+bool AccountRepositoryDB::registerStoreTransaction(
+	const uint32_t &id,
+	uint8_t type,
+	int32_t amount,
```
uint32_t

dudantas commented on src/account/account_repository_db.hpp:
```diff
@@ -32,6 +32,15 @@ class AccountRepositoryDB final : public AccountRepository {
+	bool registerStoreTransaction(
+		const uint32_t &id,
+		uint8_t type,
+		int32_t amount,
```
uint32_t

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7749,6 +7750,128 @@ void Player::closeAllExternalContainers() {
+		}
+
+		case OfferTypes_t::ALLBLESSINGS: {
+			for (uint8_t bless = 1; bless <= 8; ++bless) {
```
Following the example from here: https://github.com/opentibiabr/canary/commit/193939c4903bcbec1cf0d6e808422693f2e79278

We should remove this magic number by the equivalent using the magic enum
```suggestion
			for (auto bless : magic_enum::enum_values<Blessings>()) {
```

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7749,6 +7750,128 @@ void Player::closeAllExternalContainers() {
+
+		case OfferTypes_t::PREYBONUS: {
+			auto cardsAmount = offer->getOfferCount();
+			if (getPreyCards() + cardsAmount >= 50) {
```
It might be beneficial to establish a new key in the config.lua file for this purpose, thereby eliminating the magic number and enhancing organizational control.

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7749,6 +7750,128 @@ void Player::closeAllExternalContainers() {
+		case OfferTypes_t::BLESSINGS: {
+			auto blessId = offer->getOfferId();
+			if (blessId < 1 || blessId > 8) {
+				break;
```
```suggestion
			if (!magic_enum::enum_contains<Blessings>(static_cast<Blessings>(blessId))) {
				sendStoreError(StoreErrors_t::PURCHASE, "An error has occurred, please contact your administrator.");
				g_logger().error("[{}] invalid blessing id: {}, for player: {}", __METHOD_NAME__, blessId, getName());
				break;
```

dudantas commented on src/creatures/players/player.cpp:
```diff
@@ -7749,6 +7750,128 @@ void Player::closeAllExternalContainers() {
+			auto offerInstantAmount = offer->getOfferCount();
+			auto playerInstantAmount = getStorageValue(14901);
+
+			if (playerInstantAmount + offerInstantAmount >= 90) {
```
Here is the same, can be moved to config.lua

dudantas commented on src/creatures/players/player.hpp:
```diff
@@ -2655,6 +2665,41 @@ class Player final : public Creature, public Cylinder, public Bankable {
 	uint16_t getPlayerVocationEnum() const;
 
+	// Store functions
+	void openStore() {
```
It's a good practice to move the body of these functions to the .cpp file, which helps to avoid increasing the complexity of the .hpp file even further.

dudantas reviewed (commented):
I'll leave some suggestions, but I'll commit the corrections now.

dudantas committed:
fix: some review changes and reorganize game.cpp functions order

dudantas committed:
fix: missing review

dudantas commented on src/io/io_bosstiary.cpp:
```diff
@@ -196,7 +196,7 @@ void IOBosstiary::addBosstiaryKill(std::shared_ptr<Player> player, const std::sh
 	int32_t value = player->getStorageValue(STORAGEVALUE_PODIUM);
 	if (value != 1 && newBossLevel == 2) {
-		auto returnValue = g_game().addItemStoreInbox(player, ITEM_PODIUM_OF_VIGOUR);
+		auto returnValue = g_game().processHouseOffer(player, ITEM_PODIUM_OF_VIGOUR);
```
This item is earned, not purchased. Isn't it confusing that the function change says it was purchased from the store?

dudantas commented on src/game/game.cpp:
```diff
@@ -10298,6 +10307,129 @@ void Game::addPlayerUniqueLogin(std::shared_ptr<Player> player) {
+		return false;
+	}
+
+	Database &db = Database::getInstance();
```
We can use "g_database()".

dudantas committed:
fix: review

dudantas commented on src/game/game.cpp:
```diff
@@ -10298,6 +10307,129 @@ void Game::addPlayerUniqueLogin(std::shared_ptr<Player> player) {
+	return true;
+}
+
+bool Game::processNameChangeOffer(std::shared_ptr<Player> player, std::string &name) {
```
Wouldn't it be better to just pass the "name" as a copy? This will avoid having to create a copy inside the function, which should create the same overhead but is more complex.

dudantas committed:
fix: qodana linter and others

dudantas committed:
test: print directories

dudantas committed:
fix: correct getAccountId method

dudantas committed:
fix: message store error

dudantas committed:
feat: init implementation of "store history detail"

dudantas commented on src/game/game.cpp:
```diff
@@ -10409,6 +10539,403 @@ void Game::playerRewardChestCollect(uint32_t playerId, const Position &pos, uint
+		tempHistory.coinAmount = coinAmount;
+		playerReceptor->sendCoinBalance();
+		playerReceptor->setStoreHistory(tempHistory);
+	}
```
This seems redundant to me, the boolean in: std::shared_ptr<Player> playerReceptor = getPlayerByName(receptorName, true);

It already handles the offline player properly, doesn't it? I think this can be simplified.

dudantas committed:
feat: add "enum handling" to Database::getNumber function

dudantas committed:
improve: store detail

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
feat: multiple offer price

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
feat: try on

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
feat: store history detail

dudantas committed:
fix: qodana

dudantas committed:
fix: qodana reviews

murilo09 committed:
fix: change name

murilo09 committed:
fix: offers error message

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
remove log

dudantas committed:
feat: implement store detail

dudantas committed:
fix: circular inclusion

dudantas committed:
fix: enable unity build

dudantas committed:
Merge branch 'main' into gamestore-cpp

dudantas committed:
fix: compilation with unity build

dudantas committed:
Merge branch 'main' into gamestore-cpp

dudantas committed:
fix: initialize variables

murilo09 committed:
fix: compilation

murilo09 committed:
feat: show coin type on offer

murilo09 committed:
fix: possible crash

murilo09 committed:
fix: typo on access

murilo09 committed:
fix: shadowing class variable

murilo09 committed:
feat: rookgaard store options

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
some fixes and sonar recommendations

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
sonar recommendations

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
refactor and fixes

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
remove nested condition

elsongabriel commented on data/modules/scripts/gamestore/init.lua:
```diff
@@ -1,2268 +0,0 @@
-	player:updateUIExhausted()
-end
-
--- Used on cyclopedia store summary
-local function insertPlayerTransactionSummary(player, offer)
-	local id = offer.id
-	if offer.type == GameStore.OfferTypes.OFFER_TYPE_HOUSE then
-		id = offer.itemtype
-	elseif offer.type == GameStore.OfferTypes.OFFER_TYPE_BLESSINGS then
-		id = offer.blessid
-	end
-	player:createTransactionSummary(offer.type, math.max(1, offer.count or 1), id)
-end
```
we need to include this function for cyclopedia works.

elsongabriel commented on data/modules/scripts/gamestore/init.lua:
```diff
@@ -1,2268 +0,0 @@
-
-		player:makeCoinTransaction(offer, hirelingName)
-		local message = "You have successfully bought " .. hirelingName
-		player:createTransactionSummary(offer.type, 1)
```
include this call too.

elsongabriel commented on data/modules/scripts/gamestore/init.lua:
```diff
@@ -1,2268 +0,0 @@
-	end
-
-	if table.contains({ GameStore.OfferTypes.OFFER_TYPE_HOUSE, GameStore.OfferTypes.OFFER_TYPE_EXPBOOST, GameStore.OfferTypes.OFFER_TYPE_PREYBONUS, GameStore.OfferTypes.OFFER_TYPE_BLESSINGS, GameStore.OfferTypes.OFFER_TYPE_ALLBLESSINGS, GameStore.OfferTypes.OFFER_TYPE_INSTANT_REWARD_ACCESS }, offer.type) then
-		insertPlayerTransactionSummary(player, offer)
```
this too.

murilo09 replied:
We don’t need all these additional calls, simply modifying the function in C++ and call it will solve the issue. It’s quite straightforward.

elsongabriel replied:
> We don’t need all these additional calls, simply modifying the function in C++ and call it will solve the issue. It’s quite straightforward.

when I said to create these functions and their calls. I mean creating it on cpp code.

elsongabriel reviewed (changes requested):
I believe other small fixes and adjustes are necessary.

murilo09 committed:
Merge branch 'main' into gamestore-cpp

murilo09 committed:
Merge branch 'main' into gamestore-cpp

murilo09 committed:
Merge branch 'main' into gamestore-cpp

murilo09 committed:
fix tests

murilo09 committed:
fix positive value on store history

murilo09 committed:
fix subcategories

murilo09 committed:
feat: button redirects

murilo09 committed:
Merge branch 'main' into gamestore-cpp

murilo09 committed:
fix compilation

github-actions[bot] committed:
Code format - (Clang-format)

murilo09 committed:
Merge branch 'main' into gamestore-cpp

murilo09 committed:
Merge branch 'main' into gamestore-cpp

murilo09 committed:
fix compilation and revert database.hpp

murilo09 committed:
some fixes

murilo09 committed:
Merge branch 'main' into gamestore-cpp

murilo09 committed:
fixes and refactor part 1

murilo09 committed:
missing coin type

murilo09 committed:
feat: collections

murilo09 committed:
more collections

murilo09 committed:
fixes and refactor part 2

murilo09 committed:
sonar fixes

murilo09 committed:
removed unused variables

murilo09 committed:
fix: banner and collection redirect

murilo09 committed:
fix: temple teleport logic

murilo09 committed:
fix: outfit offer logic

and other fixes

murilo09 committed:
refactor: xml reader

murilo09 committed:
Merge branch 'main' into gamestore-cpp

murilo09 committed:
fix: outfit try on

murilo09 committed:
fix: home offers

murilo09 committed:
more refactorings on xml reader

murilo09 committed:
missing using enum

murilo09 committed:
fix types

dudantas committed:
fix: tests

dudantas committed:
fix: tests

murilo09 committed:
Merge branch 'main' into gamestore-cpp

murilo09 committed:
add missing migration

murilo09 committed:
fix build

powrogames commented:
I would suggest making possibility of accept both coins (transferable and non-transferable), It was requested many times so maybe it is good idea to make CoinType.Both and access 2 of coins kind

dudantas committed:
Merge branch 'main' into gamestore-cpp

dudantas committed:
Migrate gamestore detail to components and update DB

Renamed player_store_detail files to player_gamestore_detail under components. Added migration script to update the store_history table schema for the new gamestore implementation. Minor adjustments in game.cpp to support the migration and new outfit change logic.

dudantas commented:
Closing this PR in favor of https://github.com/opentibiabr/canary/pull/3738, which reorganizes the modules and maintains more support for what already exists, reduces the "surface" of change, and keeps everything organized in the old model, but modularized to facilitate compatibility.
